### PR TITLE
Update k8s tests and libraries to v1.23.0-rc.0

### PR DESCRIPTION
**Pull request by aanm**
Signed-off-by: André Martins <andre@cilium.io>

```release-note
Add support for k8s 1.23.0
```

### CODE REVIEW

**Review by nbusseneau — APPROVED** [review]
Code changes LGTM. I am making changes to Jenkins to add 1.23 and enable testing via `/test-1.23-4.9`.

Once this is actually merged or close to merging (depending if we want to test the rotated jobs before or after merging), we need to rotate the `/test` (`master` branch) Jenkins triggers and Jenkins pipelines. This entails:

- Adding Kernel 4.9 / K8s 1.23 testing
- Testing Kernel 4.19 on 1.22 instead of 1.21
- Testing Kernel 5.4 on 1.21 instead of 1.20
- Rolling back 1.21 to 4.9 kernel
- Updating the Table of Truth:tm:
- Update MLH config
- After updating triggers on Jenkins, trigger `/test` at least once so that the jobs get registered on GH, then update GH PR status marked as required.

Please ping me on Slack when merging 🚀

**Comment by nbusseneau** [comment]
/test-1.23-4.9

**Comment by aanm** [comment]
/test-1.23-4.9

File: Documentation/contributing/testing/e2e.rst [hunk 1]
```diff
@@ -137,6 +137,7 @@ The Kubernetes tests support the following Kubernetes versions:
 * 1.20
 * 1.21
 * 1.22
+* 1.23
 
 By default, the Vagrant VMs are provisioned with Kubernetes 1.21. To run with any other
```
Comment (nbusseneau) [inline Documentation/contributing/testing/e2e.rst]: Should be 1.22 after `jenkinsfiles/ginkgo-kernel.Jenkinsfile` has been changed.

**Review by nbusseneau — COMMENTED** [review]
Missed in previous review:

- Updating supported K8s versions in `Documentation/contributing/testing/e2e.rst`.
- Updating default K8s version in `jenkinsfiles/ginkgo-kernel.Jenkinsfile`.

File: Documentation/contributing/testing/e2e.rst (hunk 1 above)
Comment (aanm) [reply Documentation/contributing/testing/e2e.rst]: Fixed locally, will repushed once tests have passed

**Comment by aanm** [comment]
/test-1.23-4.9

**Comment by aanm** [comment]
/test-1.23-4.9

File: pkg/k8s/slim/k8s/apis/meta/v1beta1/types.go [hunk 2]
```diff
@@ -1,6 +1,18 @@
-// SPDX-License-Identifier: Apache-2.0
-// Copyright 2017 The Kubernetes Authors.
-// Copyright 2020-2021 Authors of Cilium
+/*
+Copyright 2017 The Kubernetes Authors.
+
```
Comment (qmonnet) [inline pkg/k8s/slim/k8s/apis/meta/v1beta1/types.go]: We lose the Cilium copyright notice on this file. Can you please confirm that this is intentional?

File: test/provision/k8s_install.sh [hunk 3]
```diff
@@ -358,6 +358,20 @@ case $K8S_VERSION in
         CONTROLLER_FEATURE_GATES="EndpointSlice=true,EndpointSliceTerminatingCondition=true"
         API_SERVER_FEATURE_GATES="EndpointSlice=true,EndpointSliceTerminatingCondition=true"
         ;;
+    "1.23")
+        # kubeadm 1.23 requires conntrack to be installed, we can remove this
+        # once we have upgrade the VM image version.
```
Comment (qmonnet) [inline test/provision/k8s_install.sh]:
```suggestion
        # once we have upgraded the VM image version.
```

**Review by qmonnet — APPROVED** [review]
Looks good, with one item to check below

File: pkg/k8s/slim/k8s/apis/meta/v1beta1/types.go (hunk 2 above)
Comment (aanm) [reply pkg/k8s/slim/k8s/apis/meta/v1beta1/types.go]: ups, it's was not intentional.

File: go.mod [hunk 4]
```diff
@@ -90,13 +90,13 @@ require (
...
-	k8s.io/api v0.22.3
-	k8s.io/apiextensions-apiserver v0.22.3
-	k8s.io/apimachinery v0.22.3
-	k8s.io/client-go v0.22.3
-	k8s.io/code-generator v0.22.3
+	k8s.io/api v0.24.0-alpha.0
+	k8s.io/apiextensions-apiserver v0.24.0-alpha.0
+	k8s.io/apimachinery v0.24.0-alpha.0
+	k8s.io/client-go v0.24.0-alpha.0
+	k8s.io/code-generator v0.24.0-alpha.0
```
Comment (rolinh) [inline go.mod:97]: I was told the `v0.23.0-rc.0` tag is not yet available so this tag is used instead and corresponds to the same sha.

**Review by rolinh — APPROVED** [review]

**Comment by aanm** [comment]
Fixed all comments mentioned by Quentin. Doesn't require a new CI run. Merging since previously it was all green

**Pull request merged**